sinks/external_test.go

```diff
@@ -129,28 +132,50 @@ func TestSetSinksRegisterAgain(t *testing.T) {
 func TestSetSinksStore(t *testing.T) {
 	as := assert.New(t)
 	s1 := &DummySink{}
-	m, err := NewExternalSinkManager([]sink_api.ExternalSink{s1})
+	c := cache.NewCache(time.Minute, time.Hour)
+	m, err := NewExternalSinkManager([]sink_api.ExternalSink{s1}, c, time.Microsecond)
 	as.Nil(err)
 	as.Equal(0, s1.StoredTimeseries)
 	as.Equal(0, s1.StoredEvents)
-	err = m.Store(source_api.AggregateData{})
-	as.Nil(err)
+	var (
+		pods       []source_api.Pod
+		containers []source_api.Container
+		events     []*cache.Event
+	)
+	f := fuzz.New().NumElements(1, 1).NilChance(0)
+	f.Fuzz(&pods)
+	f.Fuzz(&containers)
+	f.Fuzz(&events)
+	c.StorePods(pods)
+	c.StoreContainers(containers)
+	c.StoreEvents(events)
+	stopChan := m.Sync()
+	as.NotNil(stopChan)
+	time.Sleep(time.Second)
 	as.Equal(1, s1.StoredTimeseries)
 	as.Equal(1, s1.StoredEvents)
+	stopChan <- struct{}{}
 	err = m.SetSinks([]sink_api.ExternalSink{})
 	as.Nil(err)
-	err = m.Store(source_api.AggregateData{})
-	as.Nil(err)
+	stopChan = m.Sync()
+	as.NotNil(stopChan)
 	as.Equal(1, s1.StoredTimeseries)
 	as.Equal(1, s1.StoredEvents)
+
 	err = m.SetSinks([]sink_api.ExternalSink{s1})
+	as.Equal(1, s1.StoredTimeseries)
+	as.Equal(1, s1.StoredEvents)
 	as.Nil(err)
-	err = m.Store(source_api.AggregateData{})
-	as.Nil(err)
+	f.Fuzz(&pods)
+	f.Fuzz(&containers)
+	f.Fuzz(&events)
+	c.StorePods(pods)
+	c.StoreContainers(containers)
+	c.StoreEvents(events)
+	stopChan = m.Sync()
+	as.NotNil(stopChan)
+	time.Sleep(time.Second)
 	as.Equal(2, s1.StoredTimeseries)
```

Comment:
FWIW, I fixed a race condition in https://github.com/kubernetes/heapster/pull/405 - there might be another one, and if similar it could be causing something like this. Since `Store()` would return before it's actually finished.